src/nvs.rs

```diff
@@ -841,160 +826,78 @@ impl<T: NvsPartitionId> RawStorage for EspKeyValueStorage<T> {
-                                buf.as_mut_ptr() as *mut _,
-                                &mut len as *mut _,
-                            )
-                        })?;
-
-                        Ok(Some(&buf[..len]))
+        if let Some(data_type) = self.0.find_key(name)? {
+            return match data_type {
+                NvsDataType::Blob => self.0.get_blob(name, buf),
+                NvsDataType::U64 => {
+                    // U64 is always 8 bytes, so we can use a fixed-size buffer
+                    if buf.len() < 8 {
+                        return Err(EspError::from_infallible::<ESP_ERR_NVS_INVALID_LENGTH>());
                     }
+                    let mut u64value: u64 = 0;
+                    esp!(unsafe {
+                        nvs_get_u64(
+                            self.0 .1,
+                            to_cstring_arg(name)?.as_ptr(),
+                            &mut u64value as *mut _,
+                        )
+                    })?;
+                    buf.copy_from_slice(&u64value.to_le_bytes());
+                    Ok(Some(&buf[..8]))
                 }
-            }
-            err => {
-                // bail on error
-                esp!(err)?;
-
-                // u64 value was found, decode it
-                let len: u8 = (u64value & 0xff) as u8;
-
-                if buf.len() < len as _ {
-                    // Buffer not large enough
-                    return Err(EspError::from_infallible::<ESP_ERR_NVS_INVALID_LENGTH>());
-                }
-
-                u64value >>= 8;
-
-                let array: [u8; 7] = [
-                    (u64value & 0xff) as u8,
-                    ((u64value >> 8) & 0xff) as u8,
-                    ((u64value >> 16) & 0xff) as u8,
-                    ((u64value >> 24) & 0xff) as u8,
-                    ((u64value >> 32) & 0xff) as u8,
-                    ((u64value >> 40) & 0xff) as u8,
-                    ((u64value >> 48) & 0xff) as u8,
-                ];
-
-                buf[..len as usize].copy_from_slice(&array[..len as usize]);
-
-                Ok(Some(&buf[..len as usize]))
-            }
+                _ => Ok(None), // Other types are not supported in this storage
+            };
         }
+        Ok(None)
     }
 
     pub fn set_raw(&self, name: &str, buf: &[u8]) -> Result<bool, EspError> {
-        let c_key = to_cstring_arg(name)?;
-        let mut u64value: u_int64_t = 0;
-
         // start by just clearing this key
-        unsafe { nvs_erase_key(self.0 .1, c_key.as_ptr()) };
+        self.0.remove(name)?;
 
         if buf.len() < 8 {
-            for v in buf.iter().rev() {
-                u64value <<= 8;
-                u64value |= *v as u_int64_t;
-            }
-
-            u64value <<= 8;
-            u64value |= buf.len() as u_int64_t;
-
-            esp!(unsafe { nvs_set_u64(self.0 .1, c_key.as_ptr(), u64value) })?;
+            let value = u64::from_le_bytes(
```

Comment:
That won't work unless the buf is _exactly_ 8 bytes, but it will never be, because you have an if above which says `if buf.len() < 8`.

You need to restore the original `for` loop with the shifting logic.